mautic/mautic
# Reuse existing editor keep history 2

Replaces https://github.com/mautic/mautic/pull/10781 which had some unrelated commits.

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | yes
| Deprecations?                          |  no
| BC breaks? (use the c.x branch)        |  no
| Automated tests included?              |  no
| Related user documentation PR URL      |  N/A
| Related developer documentation PR URL | 
| Issue(s) addressed                     |  https://github.com/mautic/plugin-grapesjs-builder/pull/91, not a reported issue: code in the editor does not update, related to: https://github.com/mautic/mautic/pull/10707 Fixes #10895 

#### Description:

While working on https://github.com/mautic/mautic/pull/10707 we realized that we have another problem with the editor content not being refreshed when the editor is reopened. This PR fixes that. 

Along with that, it creates new possibilities:
- Make the Grapesjs editor available in the global `Mautic.builder` property
- Keep the editor "around" when closed. This has the advantage that e.g. the history (undo) is still available after closing and reopening it.
- Faster load times
- New configuration possibilities. See the README in the preset (maybe check the companion PR).

#### Steps to test this PR:

1. Open this PR on Gitpod or pull-down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a new email and open it in the builder
3. Open the code mode. Make sure the HTML is from the theme selected
4. Close the modal and the builder
5. Select another theme
6. Open the builder and the code editor
7. Make sure the HTML is from the newly selected theme (used to be the first one)
8. In addition make sure the correct panel is opened on the right-hand side: the blocks selector, not e.g. the Style Manager or a blank panel

This PR relies on the code in the preset:
https://github.com/mautic/grapesjs-preset-mautic/pull/25/

Once this code works the preset PR should be code reviewed and merged too.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10847"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
